images/build/setcap/Dockerfile

```diff
@@ -0,0 +1,25 @@
+# Copyright 2021 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+ARG BASEIMAGE
+
+FROM ${BASEIMAGE}
+
+# Required for s390x build.
```

Comment:
???
is this another issue with date qemu?

Reply:
If I remove those links I get the following error when building for s390x:
```
6 0.589 Get:1 http://security.debian.org/debian-security buster/updates InRelease [65.4 kB]                                                                                          
#6 0.590 Get:2 http://deb.debian.org/debian buster InRelease [122 kB]                                                                                                                 
#6 0.676 Get:3 http://deb.debian.org/debian buster-updates InRelease [51.9 kB]
#6 2.531 Get:4 http://security.debian.org/debian-security buster/updates/main s390x Packages [238 kB]
#6 4.357 Get:5 http://deb.debian.org/debian buster/main s390x Packages [7624 kB]
#6 6.144 Get:6 http://deb.debian.org/debian buster-updates/main s390x Packages [9508 B]
#6 8.218 Fetched 8110 kB in 8s (1062 kB/s)
#6 8.218 Reading package lists...
#6 11.46 Reading package lists...
#6 14.61 Building dependency tree...
#6 15.16 Reading state information...
#6 15.54 The following additional packages will be installed:
#6 15.54   libcap2
#6 15.54 Recommended packages:
#6 15.54   libpam-cap
#6 15.68 The following NEW packages will be installed:
#6 15.68   libcap2 libcap2-bin
#6 15.84 0 upgraded, 2 newly installed, 0 to remove and 4 not upgraded.
#6 15.84 Need to get 46.0 kB of archives.
#6 15.84 After this operation, 142 kB of additional disk space will be used.
#6 15.84 Get:1 http://deb.debian.org/debian buster/main s390x libcap2 s390x 1:2.25-2 [17.6 kB]
#6 15.85 Get:2 http://deb.debian.org/debian buster/main s390x libcap2-bin s390x 1:2.25-2 [28.5 kB]
#6 16.51 debconf: delaying package configuration, since apt-utils is not installed
#6 16.62 Fetched 46.0 kB in 0s (467 kB/s)
#6 16.67 Error while loading /usr/sbin/dpkg-split: No such file or directory
#6 16.67 Error while loading /usr/sbin/dpkg-deb: No such file or directory
#6 16.67 dpkg: error processing archive /var/cache/apt/archives/libcap2_1%3a2.25-2_s390x.deb (--unpack):
#6 16.67  dpkg-deb --control subprocess returned error exit status 1
#6 16.68 Error while loading /usr/sbin/dpkg-split: No such file or directory
#6 16.68 Error while loading /usr/sbin/dpkg-deb: No such file or directory
#6 16.68 dpkg: error processing archive /var/cache/apt/archives/libcap2-bin_1%3a2.25-2_s390x.deb (--unpack):
#6 16.68  dpkg-deb --control subprocess returned error exit status 1
#6 16.70 Errors were encountered while processing:
#6 16.70  /var/cache/apt/archives/libcap2_1%3a2.25-2_s390x.deb
#6 16.70  /var/cache/apt/archives/libcap2-bin_1%3a2.25-2_s390x.deb
#6 16.77 E: Sub-process /usr/bin/dpkg returned an error code (1)
```
I found a related issue: https://github.com/docker/buildx/issues/495

Reply:
huh, I hadn't seen this approach before versus ensuring qemu + the builder are setup, I'm not sure this makes sense but ... :this-is-fine:

Reply:
@cwsolee can you please look at why we need to jump through hoops like this?

Reply:
@vinayakankugoyal @dims @BenTheElder 

We are not able to reproduce the mentioned issue at our end, we need more details on the docker and buildx version that is getting used. We have tried building the cross images with below version and we are observing build failures;
```
$ docker buildx version
github.com/docker/buildx v0.5.1-docker 11057da37336192bfc57d81e02359ba7ba848e4a
$ docker version
Client: Docker Engine - Community
 Version:           20.10.2
 API version:       1.41
 Go version:        go1.13.15
 Git commit:        2291f61
 Built:             Mon Dec 28 16:17:32 2020
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true
Server: Docker Engine - Community
 Engine:
  Version:          20.10.2
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.13.15
  Git commit:       8891c58
  Built:            Mon Dec 28 16:15:09 2020
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.4.3
  GitCommit:        269548fa27e0089a8b8278fc4fc781d7f65a939b
 runc:
  Version:          1.0.0-rc92
  GitCommit:        ff819c7e9184c13b7c2607fe6c30ae19403a7aff
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```
We get below error while building cross image for s390x/arm64/ppc64le:

```
[5:49 AM] docker buildx build \
        --pull \
        --load \
        --platform linux/s390x \
        -t "gcr.io/k8s-staging-build-image"/setcap-s390x:buster-v1.4.0 \
        -t "gcr.io/k8s-staging-build-image"/setcap-s390x:v0.7.0-64-g6bd12162-buster \
        -t "gcr.io/k8s-staging-build-image"/setcap-s390x:latest-buster \
        --build-arg=BASEIMAGE=k8s.gcr.io/build-image/debian-base-s390x:buster-v1.4.0 \
        .
[+] Building 21.2s (10/10) FINISHED
 => [internal] booting buildkit                                                                                                                                    3.8s
 => => pulling image moby/buildkit:buildx-stable-1                                                                                                                 1.4s
 => => creating container buildx_buildkit_interesting_nightingale0                                                                                                 2.4s
 => [internal] load build definition from Dockerfile                                                                                                               0.1s
 => => transferring dockerfile: 882B                                                                                                                               0.0s
 => [internal] load .dockerignore                                                                                                                                  0.1s
 => => transferring context: 2B                                                                                                                                    0.0s
 => [internal] load metadata for k8s.gcr.io/build-image/debian-base-s390x:buster-v1.4.0                                                                            1.1s
 => [1/6] FROM k8s.gcr.io/build-image/debian-base-s390x:buster-v1.4.0@sha256:610526b047d4b528d9e14b4f15347aa4e37af0c47e1307a2f7aebf8745c8a323                      2.9s
 => => resolve k8s.gcr.io/build-image/debian-base-s390x:buster-v1.4.0@sha256:610526b047d4b528d9e14b4f15347aa4e37af0c47e1307a2f7aebf8745c8a323                      0.0s
 => => sha256:376d61ddc98ab78b241628e7663b118ecccbc700046029c73e2a333219d21bce 23.06MB / 23.06MB                                                                   1.6s
 => => extracting sha256:376d61ddc98ab78b241628e7663b118ecccbc700046029c73e2a333219d21bce                                                                          1.3s
 => [2/6] RUN ln -s /usr/bin/dpkg-split /usr/sbin/dpkg-split                                                                                                       0.5s
 => [3/6] RUN ln -s /usr/bin/dpkg-deb /usr/sbin/dpkg-deb                                                                                                           0.4s
 => [4/6] RUN ln -s /bin/tar /usr/sbin/tar                                                                                                                         0.4s
 => [5/6] RUN ln -s /bin/rm /usr/sbin/rm                                                                                                                           0.5s
 => ERROR [6/6] RUN clean-install libcap2-bin                                                                                                                      0.4s
------
 > [6/6] RUN clean-install libcap2-bin:
#10 0.303 /bin/sh: 0: Can't open clean-install
------
Dockerfile:25
--------------------
  23 |     RUN ln -s /bin/rm /usr/sbin/rm
  24 |
  25 | >>> RUN clean-install libcap2-bin
  26 |
--------------------
error: failed to solve: rpc error: code = Unknown desc = executor failed running [/dev/.buildkit_qemu_emulator /bin/sh -c clean-install libcap2-bin]: exit code: 127
Makefile:43: recipe for target 'build' failed
make[1]: *** [build] Error 1
make[1]: Leaving directory '/home/gary/release/images/build/setcap'
Makefile:67: recipe for target 'sub-build-s390x' failed
make: *** [sub-build-s390x] Error 2
```

Although we can perform the step of `clean-install libcap2-bin`  into a `k8s.gcr.io/build-image/debian-base-s390x:buster-v1.4.0` container run on amd64 machine by registering QEMU manually without performing below steps:
```RUN ln -s /usr/bin/dpkg-split /usr/sbin/dpkg-split
RUN ln -s /usr/bin/dpkg-deb /usr/sbin/dpkg-deb
RUN ln -s /bin/tar /usr/sbin/tar
RUN ln -s /bin/rm /usr/sbin/rm
```
For now we suspect on the buildx version being used and not on QEMU, need more details on docker and buildx version used to verify/reproduce issue at our end.

/cc @cwsolee

Reply:
I updated docker and buildx and it seems to be working now!
```
$ docker buildx version 
github.com/docker/buildx v0.5.1-docker 11057da37336192bfc57d81e02359ba7ba848e4a
```
```
$ docker version
Client: Docker Engine - Community
 Version:           20.10.2
 API version:       1.41
 Go version:        go1.13.15
 Git commit:        2291f61
 Built:             Mon Dec 28 16:17:34 2020
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.2
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.13.15
  Git commit:       8891c58
  Built:            Mon Dec 28 16:15:28 2020
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.4.3
  GitCommit:        269548fa27e0089a8b8278fc4fc781d7f65a939b
 runc:
  Version:          1.0.0-rc92
  GitCommit:        ff819c7e9184c13b7c2607fe6c30ae19403a7aff
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```
clean-install is failing as you mentioned but when I replace `RUN clean-install libcap2-bin` with:
```
RUN apt-get update && apt-get -y --no-install-recommends install libcap2-bin
```
it works!

Reply:
Ok so I got `RUN clean-install libcap2-bin` to work
I added `RUN chmod +x /usr/local/bin/clean-install` before `RUN clean-install libcap2-bin`. `make sub-build-s390x` passed, very weird 😕

Reply:
Whats weirder is that:
```
$ docker run --rm k8s.gcr.io/build-image/debian-base-s390x:buster-v1.4.0 ls -l /usr/local/bin/clean-install 
WARNING: The requested image's platform (linux/s390x) does not match the detected host platform (linux/amd64) and no specific platform was requested
-rwxr-xr-x 1 root root 1050 Jan 27 11:12 /usr/local/bin/clean-install
```

Reply:
> I added RUN chmod +x /usr/local/bin/clean-install before RUN clean-install libcap2-bin in the Dockerfile. make sub-build-s390x passed, very weird 😕

AHA! this repo has a bug then, we fixed this in the k/k build previously.
The reason for this is likely your local `umask`, which we should not be relying on.

Reply:
- https://github.com/kubernetes/kubernetes/pull/90994
- https://github.com/kubernetes/kubernetes/pull/90779
- https://github.com/kubernetes/kubernetes/pull/67222